src/app/loadout-builder/hooks/useProcess.ts

```diff
@@ -0,0 +1,209 @@
+import { wrap, releaseProxy } from 'comlink';
+import { useEffect, useState, useMemo } from 'react';
+import { ItemsByBucket, LockedMap, LockedArmor2ModMap, ArmorSet, StatTypes } from '../types';
+import { DimItem, DimSocket, DimSockets, D2Item } from 'app/inventory/item-types';
+import {
+  ProcessItemsByBucket,
+  ProcessItem,
+  ProcessArmorSet,
+  ProcessSocket,
+  ProcessSockets,
+} from '../processWorker/types';
+
+interface ProcessState {
+  processing: boolean;
+  result: {
+    sets: ArmorSet[];
+    combos: number;
+    combosWithoutCaps: number;
+  } | null;
+  currentCleanup: (() => void) | null;
+}
+
+type ItemsById = { [id: string]: DimItem };
+
+/**
+ * Hook to process all the stat groups for LO in a web worker.
+ */
+export function useProcess(
+  filteredItems: ItemsByBucket,
+  lockedItems: LockedMap,
+  lockedArmor2ModMap: LockedArmor2ModMap,
+  assumeMasterwork: boolean,
+  statOrder: StatTypes[]
+) {
+  const [{ result, processing, currentCleanup }, setState] = useState({
+    processing: false,
+    result: null,
+    currentCleanup: null,
+  } as ProcessState);
+
+  const { worker, cleanup } = useWorkerAndCleanup(
+    filteredItems,
+    lockedItems,
+    lockedArmor2ModMap,
+    assumeMasterwork,
+    statOrder
+  );
+
+  if (currentCleanup && currentCleanup !== cleanup) {
+    currentCleanup();
+  }
+
+  useEffect(() => {
+    const processStart = performance.now();
+
+    setState({ processing: true, result, currentCleanup: cleanup });
+
+    const processItems: ProcessItemsByBucket = {};
+    const itemsById: ItemsById = {};
+
+    for (const [key, items] of Object.entries(filteredItems)) {
+      processItems[key] = [];
+      for (const item of items) {
+        if (item.isDestiny2()) {
+          processItems[key].push(mapDimItemToProcessItem(item));
+          itemsById[item.id] = item;
+        }
+      }
+    }
+
+    const workerStart = performance.now();
+    worker
+      .process(processItems, lockedItems, lockedArmor2ModMap, assumeMasterwork, statOrder)
+      .then(({ sets, combos, combosWithoutCaps }) => {
+        console.log(`useProcess: worker time ${performance.now() - workerStart}ms`);
+        const hydratedSets = sets.map((set) => hydrateArmorSet(set, itemsById));
+
+        setState({
+          processing: false,
+          result: {
+            sets: hydratedSets,
+            combos,
+            combosWithoutCaps,
+          },
+          currentCleanup: null,
+        });
+
+        console.log(`useProcess ${performance.now() - processStart}ms`);
```

Comment:
You can also use `console.time('useProcess')` and `console.timeEnd('useProcess')`.

Reply:
I did have that but you get a bunch of warnings if a worker gets terminated early. Not sure if we really care about that so I can do whatever you prefer.

Reply:
Gotcha. We can leave it then.